### PR TITLE
Fix DuplicateSignatures caused by races on frozen banks

### PR DESCRIPTION
**Pull request by carllin**
#### Problem
When bank.new_from_parent() is called, there is no guarantee that all recorded transactions on that bank have been committed. It's thus possible that the new child bank records a transaction with a duplicate signature as a transaction that is pending commitment on the parent bank.

#### Summary of Changes
Clone an Arc of all parent bank locks into the child bank so that the child does not attempt to operate on accounts that are still uncommitted in the parent/grandparents.

Fixes #

### CODE REVIEW

**Comment by rob-solana** [comment]
there's a change to bank.rs, too, right?

File: runtime/src/accounts.rs [hunk 1]
```diff
@@ -292,20 +322,37 @@ impl Accounts {
...
-        locks: &mut HashSet<Pubkey>,
+        (fork_locks, parent_locks): &mut AccountLocks,
         keys: &[Pubkey],
         error_counters: &mut ErrorCounters,
     ) -> Result<()> {
         // Copy all the accounts
+        let mut fork_locks = fork_locks.lock().unwrap();
         for k in keys {
-            if locks.contains(k) {
+            let mut is_locked = false;
```
Comment (rob-solana) [inline runtime/src/accounts.rs]: this is clever

File: runtime/src/accounts.rs [hunk 2]
```diff
@@ -292,20 +322,40 @@ impl Accounts {
...
-            if locks.contains(k) {
+            let is_locked = {
+                if fork_locks.contains(k) {
+                    true
+                } else {
+                    // Check parent locks. As soon as a set of parent locks is empty,
+                    // we can remove it from the list b/c that means the parent has
+                    // released the locks.
+                    let mut is_locked = false;
+                    parent_locks.retain(|p| {
```
Comment (sakridge) [inline runtime/src/accounts.rs]: this looks a lot like the really slow load_account code..

Comment (carllin) [reply runtime/src/accounts.rs]: It should be faster in that rarely should there be more than one parent (when a parent contains accounts that have not yet been unlocked on a slot boundary), and when that parent's locks are cleaned up, the parent is removed from the child.

**Pull request merged**